src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp

```diff
@@ -24,155 +24,103 @@ using namespace opossum;  // NOLINT
  *
  * @returns   Boolean value denoting whether at the group-by list of @param aggregate_node changed.
  */
-bool reduce_group_by_columns_for_constraint(const TableConstraintDefinition& table_constraint,
-                                            const std::set<ColumnID>& group_by_columns,
-                                            const std::shared_ptr<const StoredTableNode>& stored_table_node,
-                                            AggregateNode& aggregate_node) {
+bool reduce_group_by_columns_for_fd(const FunctionalDependency& fd, const ExpressionUnorderedSet& group_by_columns,
+                                    AggregateNode& aggregate_node) {
   auto group_by_list_changed = false;
-  const auto& constraint_columns = table_constraint.columns;
 
-  // Intersect primary key/unique columns and group-by columns. In case a primary key/unique constraint covers
-  // multiple columns, we need to check that all columns are present in order to later remove dependent columns.
-  std::vector<ColumnID> intersection;
-  std::set_intersection(constraint_columns.begin(), constraint_columns.end(), group_by_columns.begin(),
-                        group_by_columns.end(), std::back_inserter(intersection));
-
-  // Skip the current constraint as the primary key/unique constraint is not completely present.
-  if (intersection.size() != constraint_columns.size()) {
+  // To benefit from this rule, the FD's columns have to be part of the group-by list
+  if (!std::all_of(fd.first.cbegin(), fd.first.cend(),
+                   [&group_by_columns](std::shared_ptr<AbstractExpression> constraint_col_expr) {
+                     return group_by_columns.contains(constraint_col_expr);
+                   })) {
     return false;
   }
 
-  // Every column that is part of the table but not of the primary key/unique constraint is going to be moved from
-  // the group-by list to the list of aggregates wrapped in an ANY().
+  // Every column that is functionally dependent gets moved from
+  // the group-by list to the aggregate list. For this purpose it is wrapped in an ANY() expression.
   for (const auto& group_by_column : group_by_columns) {
-    if (std::find(constraint_columns.begin(), constraint_columns.end(), group_by_column) != constraint_columns.end()) {
-      // Do not touch primary key/unique constraint columns.
-      continue;
+    DebugAssert(std::dynamic_pointer_cast<LQPColumnExpression>(group_by_column), "Unexpected group-by-column type.");
```

Comment:
This breaks the following:

```
generate_tpch .01
SELECT p_partkey, p_partkey + 1 FROM part GROUP BY p_partkey, p_partkey + 1
```

Reply:
Could you explain the problems and what this assertion is supposed to guard against?

Even an aggregate expression could be a group by expression, e.g.:

`SELECT mb, COUNT(*) FROM (SELECT c, MIN(b) as mb FROM mixed GROUP BY c) t GROUP BY mb`

Reply:
> Could you explain the problems and what this assertion is supposed to guard against?

Some time ago, I experienced an error in the code above because of an aggregate column that slipped into the group by columns set. 
The root of the problem was an index error in the code that collected the group-by-columns.

I removed the DebugAssert.